### DISCUSSION TITLE
drawing round rects on canvas

### DISCUSSION THREAD

**Discussion by robertoranon** [thi-ng/umbrella#433] [Q&A] [answered]
Hi, 
  I have a bunch of `Rect` objects that I draw on an HTML Canvas using the `draw` function from the hiccup-canvas package. How can I draw them with round corners? The docs in hiccup-canvas mention a radius parameter, but I don't understand how to set it in the rects. 
 Thanks!

**Comment by postspectacular** [comment, marked as answer]
Hi @robertoranon - so far rounded rects were in hiccup-canvas only directly supported via vanilla hiccup shape syntax, but I've just [updated the `draw()` function in this package](https://github.com/thi-ng/umbrella/commit/573245eb07f2849df0e4e463cd9e5c40bfd37a2d) to also support thi.ng/geom shapes. For those shape you just use the optional `r` attribute to define the corner radius (which defaults to 0).

```ts
// a rounded rect with radius=20
rect([0,0], [100, 200], { r: 20 })
```

Just know that this attrib is **only** used for canvas drawing and/or SVG conversion (just updated thi.ng/hiccup-svg to support this `r` attrib too) - but the attrib has no impact on any other shape calculations/transformations in thi.ng/geom. If you really do need the latter, you'll need to (manually for now) construct a rounded rect via thi.ng/geom's `pathBuilder()` or similar (e.g. via `asCubic()` and `pathFromCubics()`).

```ts
pathFromCubics(asCubic(rect([100,200]), { breakPoints: true, uniform: false, scale: 0.25 }))
```

Hth!

**Reply by cdaein** [reply]
Would you have an example of using vanilla hiccup shape syntax for canvas drawing (with `hiccup-canvas`)?

**Reply by postspectacular** [reply]
Sure. First off, I've been using this opportunity to clean up and extract some of the existing canvas creation & adaptation functions and have just added a new package [thi.ng/canvas](https://thi.ng/canvas) (which also deprecates [thi.ng/adapt-dpi](https://thi.ng/adapt-dpi)).

As part of doing this, there's now also a small "vanilla" hiccup-canvas demo here:

Demo (refresh for random combo):
https://demo.thi.ng/umbrella/hiccup-canvas-basics/

Source:
https://github.com/thi-ng/umbrella/blob/develop/examples/hiccup-canvas-basics/

Btw. All supported shape types and their syntax are listed in the package readme:
https://github.com/thi-ng/umbrella/tree/develop/packages/hiccup-canvas#supported-shape-types

Hope that helps!

**Reply by cdaein** [reply]
This is immensely helpful. Thank you!

**Reply by postspectacular** [reply]
Great! :) The import thing to remember is that you can choose from 3 levels of abstraction:

1. Using thi.ng/geom shapes (or groups/hierarchies), which implement the [`IToHiccup` interface](https://docs.thi.ng/umbrella/api/interfaces/IToHiccup.html)

For example for the `Rect` type:
https://github.com/thi-ng/umbrella/blob/36d93175f8fbc0190c3c96a9ac0016a9663e9a80/packages/geom/src/api/rect.ts#L48-L50

2. Using the thi.ng/hiccup array convention directly (as shown above) with the `draw()` function, or...
3. Directly using shape-specific draw functions in hiccup-canvas, e.g.

https://github.com/thi-ng/umbrella/blob/36d93175f8fbc0190c3c96a9ac0016a9663e9a80/packages/hiccup-canvas/src/rect.ts#L5-L38

This last option is generally not really recommended, since these functions are considered "low level" and don't take care of any canvas state handling as the other two routes (i.e. applying, storing and restoring stroke/fill, transforms, etc.). It all depends on your use case... 😉

**Comment by robertoranon** [comment]
wow, so fast,  thanks!

![areaLights-oojw8WaJcL2EKdCknFu4ihk5T8sQSoxpqqKxRBQ1sYTqBVMbJ2m-20231219-124457](https://github.com/thi-ng/umbrella/assets/6480367/bceeb850-2cd6-41f1-a448-068a1df5030f)

**Reply by postspectacular** [reply]
WOW!! Beautiful system! :)

Btw. There's a little bugfix still incoming. The new `draw()` function currently breaks if a shape has NO attribs. New version should be ready in ~5 mins...
